src/AI/AIUtils.java

```diff
@@ -0,0 +1,169 @@
+package AI;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import CommandingOfficers.Commander;
+import Engine.GameAction;
+import Engine.GameAction.ActionType;
+import Engine.GameActionSet;
+import Engine.Path;
+import Engine.Utils;
+import Engine.XYCoord;
+import Terrain.GameMap;
+import Terrain.Location;
+import Units.Unit;
+
+public class AIUtils
+{
+  /**
+   * Figure out all actions unit can perform this turn, and organizes them by location.
+   * @param unit The unit under consideration.
+   * @param gameMap The world in which the Unit lives.
+   * @return a Map of XYCoord to ArrayList<GameActionSet>. Each XYCoord will have a GameActionSet for
+   * each type of action the unit can perform from that location.
+   */
+  public static Map<XYCoord, ArrayList<GameActionSet> > getAvailableUnitActions(Unit unit, GameMap gameMap)
+  {
+    Map<XYCoord, ArrayList<GameActionSet> > actions = new HashMap<XYCoord, ArrayList<GameActionSet> >();
+
+    // Find the possible destinations.
+    ArrayList<XYCoord> destinations = Utils.findPossibleDestinations(unit, gameMap);
+
+    for( XYCoord coord : destinations )
+    {
+      // Figure out how to get here.
+      Path movePath = Utils.findShortestPath(unit, coord, gameMap);
+
+      // Figure out what I can do here.
+      ArrayList<GameActionSet> actionSets = unit.getPossibleActions(gameMap, movePath);
+
+      // Add it to my collection.
+      actions.put(coord, actionSets);
+    }
+
+    return actions;
+  }
+
+  /**
+   * Finds all actions available to unit, and organizes them by type instead of by location.
+   * @param unit The unit under consideration.
+   * @param gameMap The world in which the Unit lives.
+   * @return a Map of ActionType to ArrayList<GameAction>.
+   */
+  public static Map<ActionType, ArrayList<GameAction> > getAvailableUnitActionsByType(Unit unit, GameMap gameMap)
+  {
+    // Create the ActionType-indexed map, and ensure we don't have any null pointers.
+    Map<ActionType, ArrayList<GameAction> > actionsByType = new HashMap<ActionType, ArrayList<GameAction> >();
+    for( ActionType atype : ActionType.values() )
+    {
+      actionsByType.put(atype, new ArrayList<GameAction>());
+    }
+
+    // First collect the actions by location.
+    Map<XYCoord, ArrayList<GameActionSet> > actionsByLoc = getAvailableUnitActions(unit, gameMap);
+
+    // Now re-map them by type, irrespective of location.
+    for( Entry<XYCoord, ArrayList<GameActionSet>> entry : actionsByLoc.entrySet() )
```

Comment:
You could just use the valueSet()s since it you aren't using the key at all.
Isn't a big deal, though.

Reply:
True.